### PR TITLE
Adds a skeleton recursive decent parser.

### PR DESCRIPTION
**Pull request by tylercecil**
Implements the parser, which currently only rejects or accepts strings. Does not make an attempt to return useful exceptions, or the eventual AST (however it should be clear how that will be added).

The PR also includes a list of good and bad strings in the tests — would certainly welcome suggested test additions.

In future PRs this will actually return an AST, but for incremental sake here is something intermediate.

### CODE REVIEW

File: src/parser.ts [hunk 1]
```diff
@@ -15,13 +15,119 @@ type TokenKind =
   | 'WHITESPACE'
   | 'EOF';
 
+/**
+ * Parses a given string, returning a Tree object.
+ *
+ * @param input
+ * @returns Resulting tree.
```
Comment (duncgibbs) [inline src/parser.ts:22]: I would make a `ChomTree` type.

File: src/parser.ts [hunk 2]
```diff
@@ -15,13 +15,119 @@ type TokenKind =
...
+ */
+function parse(input: string) {
```
Comment (duncgibbs) [inline src/parser.ts:24]: I would also return the type (`ChomTree`).

This actually goes for all of your functions.

Comment (duncgibbs) [inline src/parser.ts:24]: Structurally, I would call this file `index` or `chomTree` and then move the parser functions to a file. Actually after typing it out and really considering it, maybe not. Do what you want.

Comment (tylercecil) [reply src/parser.ts:24]: What do you mean? Isn't this file *only* parser functions? Meaning... I've already done that (except that `index` doesn't exist yet).

Comment (tylercecil) [reply src/parser.ts:24]: We talked offline about this: there was confusion about the sub functions of parse being a closures. Resolved.

**Pull request merged**